### ISSUE TITLE
addition_rnn example

### ISSUE THREAD

**Issue by lemuriandezapada** [fchollet/keras#698]
in the addition_rnn example shouldn't one use TimeDistributedDense instead of Dense for the last layer, since you have return_sequences=True?

Otherwise it doesn't really add up somehow

**Comment by Smerity** [comment]
Hi @lemuriandezapada :)

By "doesn't really add up somehow", do you mean it doesn't get the right answer for addition? I believe it's been independently checked to work by a number of people at this point. It can take a large number of iterations to hit 99% accuracy however, especially when only given a small number of training samples.

Using `TimeDistributedDense` would equate to a different task. The papers that the `addition_rnn` are based upon use the sequence to sequence architecture, which enables an arbitrary length sequence as output. By having a `Dense` layer at the end, the RNN needs to produce the correct digit at the right timestep, and can be called upon as many times a required.

If one used `TimeDistributedDense`, you'd be restricted to a given output size. The papers in fact feed in the output from the previous timestep as input to the current timestep, something that wouldn't be possible with `TimeDistributedDense`.

**Comment by Smerity** [comment]
Hi @lemuriandezapada :)

By "doesn't really add up somehow", do you mean it doesn't get the right answer for addition? I believe it's been independently checked to work by a number of people at this point. It can take a large number of iterations to hit 99% accuracy however, especially when only given a small number of training samples.

I'll preface this with - whilst I wrote the code initially, I assumed my understanding of `TimeDistributedDense` was correct. I assumed that it's a dense layer over the entirety of the previous sequence - vaguely like a Convolution1D where the filter size would cover the entirety of the sequence for each element. I might well be wrong as the naming of `TimeDistributedDense` seems (complicated)[https://github.com/fchollet/keras/issues/46]. I've just run an experiment with both and `TimeDistributedDense` takes longer to perform as well as `Dense` and is also slower to train (I presume as the `TimeDistributedDense` layer would be larger).

Assuming the above is correct, using `TimeDistributedDense` would equate to a different task. The papers that the `addition_rnn` are based upon use the sequence to sequence architecture, which enables an arbitrary length sequence as output. By having a `Dense` layer at the end, the RNN needs to produce the correct digit at the right timestep, and can be called upon as many times a required.

If one used `TimeDistributedDense`, you'd be restricted to a given output size. The papers in fact feed in the output from the previous timestep as input to the current timestep, something that wouldn't be possible with `TimeDistributedDense`.

**Comment by lemuriandezapada** [comment]
The results might just as well work, I haven't actually checked it. I just read the code and didn't make sense in my head :P
TimeDistributedDense applies the Dense for every time step... so the same weight matrix applied at every timestep.
Dense would just make one big weight matrix (?) over the entire state space or something similar. How do you feed the outputs back in your case? I see nothing of the like. Just one long direct predict and output collection.

**Comment by Smerity** [comment]
`Dense` and `TimeDistributedDense` are almost equivalent apart from a `dimshuffle` (see [get_output](https://github.com/fchollet/keras/blob/e379fff4258a39f3617df256d2a50adc301109ef/keras/layers/core.py#L509)). The only functional difference I've found is that `TimeDistributedDense` can be used as the first layer after input whilst `Dense` cannot.

Running [test code I wrote](https://gist.github.com/Smerity/39ffcc67f82d42faef84) results in the same training + testing loss and weights for both layer types.

I'd be interested in potentially reopening https://github.com/fchollet/keras/issues/278 as it seems it's still an active issue.

**Comment by lemuriandezapada** [comment]
Hm, fascinating. Maybe I was just using the things wrong then. Wouldn't that dimension shuffle actually have an effect on the computation?

**Comment by Smerity** [comment]
Sadly I don't have time to investigate the internals properly. I'd have certainly thought that to be the case but tested in [the script above](https://gist.github.com/Smerity/39ffcc67f82d42faef84) to ensure it worked for input / output weight matrices that aren't square. Increasing the input / output difference from 1 to 100 results in the exact same bias and almost equivalent weights. The difference in the weights is small enough to be floating point discrepancies, especially as the minor discrepancies increase with the number of epochs.

Hopefully someone with better understanding and insights will be able to come in as I won't have the time to properly investigate for a while.

**Comment by lemuriandezapada** [comment]
Ah ok, cool, thanks.
I'll dig through it further :)
